### DISCUSSION TITLE
Finding the Lowest Energy Stacking Configuration for Two Layers with Specific Miller Indices Using Pymatgen.

### DISCUSSION THREAD

**Discussion by hongyi-zhao** [materialsproject/pymatgen#3665] [Q&A] [answered]
Dear Pymatgen Development Team,

I am currently exploring the theoretical aspects of heterostructures formed by two different single layers, each defined by specific Miller indices. A pivotal component of my study is identifying the stacking configuration between these layers that manifests the lowest energy, thereby elucidating the consequential electronic and structural traits.

In my quest to achieve this objective, I have come across several methodologies that appear promising for implementation within my research framework:

- Computing Cell of Non-identical Displacement (CNID), which is a sublattice of [Displacement Shift Complete (DSC)](https://www.tf.uni-kiel.de/matwis/amat/def_en/kap_7/backbone/r7_1_3.html): This approach, implemented in [interface_master](https://github.com/nmdl-mizo/interface_master), focuses on evaluating different displacements to ascertain the optimal alignment of two layers constituting the interface.

- Wyckoff Position-Based Method: Implemented in [Hetero2d](https://github.com/cmdlab/Hetero2d), this method leverages Wyckoff positions to facilitate the search for the energetically most favorable stacking configurations between two-dimensional materials.

Given the aforementioned methods and Pymatgen's extensive capabilities in manipulating and analyzing crystal structures, I am reaching out to inquire if Pymatgen encompasses functionalities akin to these approaches, or if there exist recommended workflows that could seamlessly integrate such methodologies for efficiently identifying the lowest energy stacking configuration between two specified layers.

I look forward to your guidance and suggestions.

Best Regards,
Zhao

**Comment by wladerer** [comment, marked as answer]
This is better for the discussion section of pymatgen, but I will leave some resources here and maybe someone (or yourself) could close this

The pymatgen [analysis.interfaces](https://pymatgen.org/pymatgen.analysis.interfaces.html) module could help you with input generation and structure matching (maybe), but I doubt there are any routines built in that will do everything you are looking for

This paper could be of use to you
https://joss.theoj.org/papers/10.21105/joss.05886

And slightly less relevant, but still a nice paper
https://www.sciencedirect.com/science/article/abs/pii/S0022311523003690#bib0028

You may also look at DFT-CP by the Frederickson group. They have some interesting insights into how "chemical pressure" influences interlayer stacking
https://pubs.acs.org/doi/pdf/10.1021/ja300685j

Best of luck

**Reply by janosh** [reply]
thanks for giving some pointers @wladerer. 👍

i moved this to discussion since it's a not an actionable feature request nor code issue.
